### PR TITLE
Issue340 tolerance

### LINKED ISSUE #363: Rounding errors in final time give warning in BuildingsPy
Running `../bin/runUnitTests.py -t optimica` gives output that includes
```
*** Warning: While processing file Buildings_Fluid_Geothermal_Borefields_BaseClasses_HeatTransfer_ThermalResponseFactors_Validation_GFunction_100boreholes_result.mat for variable gFun_int: different end time between reference and test data.
tNew = [0.0, 370000000000.0]
tOld = [0.0, 370000003072.0]

*** Warning: While processing file Buildings_Fluid_Geothermal_Borefields_BaseClasses_HeatTransfer_ThermalResponseFactors_Validation_GFunction_SmallScaleValidation_result.mat for variable gFun_int: different end time between reference and test data.
tNew = [0.0, 1800000000000.0]
tOld = [0.0, 1799999979520.0]
```
This issue is to avoid such rounding errors.

### PR DESCRIPTION
**Pull request by AntoineGautier**
This closes #363 and partially addresses #340 
A fix preventing a browser instance in case of en empty plot was also implemented.
CI testing with this feature branch is run at https://github.com/lbl-srg/modelica-buildings/runs/1456557219

### CODE REVIEW

**Comment by mwetter** [comment]
@AntoineGautier : Why would the following change (to the master) not be sufficient to enable a test for relative tolerance? In view of https://github.com/lbl-srg/funnel/blob/master/src/tubeSize.c#L148-L158, this should suffice?
```
@@ -296,6 +296,7 @@ class Tester(object):
                 tol, numbers.Real):  # if scalar, considered as absolute tolerance value for x an y
             self._tol['ax'] = tol
             self._tol['ay'] = tol
+            self._tol['ry'] = tol
         elif isinstance(tol, dict):
             self._tol = tol
         else:

```

**Comment by AntoineGautier** [comment]
@mwetter I agree, that is a better solution than the one implemented on this branch. 
The issue that `y = max(tol.atoly, tol.rtoly * rangeY)` still equals to `tol.atoly` for variables with a high order or magnitude but a low variation range (for instance, absolute pressure or enthalpy) remains though.
But I think we can move forward with what you propose for now, that should at least solve the typical tolerance violation for energy consumption.
Let me implement what you propose in this branch (which contains another useful fix) and reopen that PR.

**Comment by AntoineGautier** [comment]
@mwetter In addition to the limitation mentioned in https://github.com/lbl-srg/BuildingsPy/pull/396#issuecomment-800608946 (for variables with a high order of magnitude but a low range), there is another issue for variables with a high order of magnitude, which can also take low values. Take for instance a variable varying within [0, 1E10], with both an absolute tolerance and relative tolerance of 1E-3 (the default), all values (including near 0) will be compared against an actual absolute tolerance of max(1E-3, 1E-3*1E10)=1E7. This is problematic and a no-go in my opinion.
I think what we need to properly address this issue is a third tolerance type in funnel, namely a local (or "really relative") tolerance that scales with each element of the array of reference results.
Or change the meaning of the current relative tolerance since I don't see where it comes handy, and is rather misleading in most of the cases. That change should probably happen in https://github.com/modelica-tools/csv-compare though.

**Comment by mwetter** [comment]
@AntoineGautier : I think in tubeSize.c, the assignment `tubeSize[1] = y;` should be an array, with the width for each point being `y[i] = max(tol.atoly, tol.rtoly * abs(y[i]))`

**Comment by AntoineGautier** [comment]
The development for integrating a local relative tolerance is done at https://github.com/lbl-srg/funnel/pull/50.

**Comment by AntoineGautier** [comment]
This is ready to merge when all tests pass.
(The changes were also tested on https://github.com/lbl-srg/modelica-buildings/pull/2414.)

File: buildingspy/development/regressiontest.py [hunk 1]
```diff
@@ -2517,7 +2538,7 @@ def _checkReferencePoints(self, ans):
                         get_user_prompt = False
 
                 except UnicodeDecodeError as e:
-                    em = "UnicodeDecodeError({0}): {1}".format(e.errno, e)
+                    em = "UnicodeDecodeError: {1}".format(e)
```
Comment (mwetter) [inline buildingspy/development/regressiontest.py:2541]: This should be `UnicodeDecodeError: {0}`, not `1`

**Comment by mwetter** [comment]
@AntoineGautier : This is good to merge when all tests pass. Please see my commit https://github.com/lbl-srg/BuildingsPy/pull/396/commits/1ac5a8042068c1daeabec27a4bfe247b31e325fd to correct a print statement.

**Pull request merged**